File: AlogoExpert-OthersLang/0623middleNode.py
# O(N) time | O(1) space
# def middleNode(linkedList):
#     count =0
#     currentNode = linkedList
#     while currentNode is not None:
#         currentNode = currentNode.next
#         count+=1
# middleNode moving with the currentNode moving    
#     middleNode = linkedList
#     for i in range(count//2):
#         middleNode = middleNode.next
#     return middleNode

# second solution
# O(N) time | O(1) space
def middleNode(linkedList):
    slowNode = linkedList
    fastNode = linkedList
    while fastNode is not None and fastNode.next is not None:
        slowNode = slowNode.next
        fastNode = fastNode.next.next # we do know it won't throw an error, we know that fast node.next is not None. So fastNode.next.next is could potentially be None, but it won't throw an error.
    return slowNode # when I reach the end with the fast node, the slow node will be at the middle node.

File: AlogoExpert-Done-Q/0623middleNode.ts
export class LinkedList {
  value: number;
  next: LinkedList | null;

  constructor(value: number) {
    this.value = value;
    this.next = null;
  }
}

// export function middleNode(linkedList: LinkedList) {
//   let slow: LinkedList | null = linkedList;
//   let fast: LinkedList | null = linkedList;
//   while (fast !== null && fast.next !== null) {
//     slow = slow!.next;
//     fast = fast.next.next;
//   }
//   return slow;
// }

// O(n) time | O(1) space
//second solution

export function middleNode(linkedList: LinkedList) {
  const length = getLength(linkedList);
  let i = 0;
  const middleNode = Math.floor(length / 2);
  let currentNode: LinkedList | null = linkedList;
  while (i < middleNode) {
    currentNode = currentNode!.next;
    i++;
  }
  return currentNode;
}

function getLength(linkedList: LinkedList) {
  let length = 0;
  let currentNode: LinkedList | null = linkedList;
  while (currentNode !== null) {
    length++;
    currentNode = currentNode!.next;
  }
  return length;
}
